x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+	askToRemove, _ := exchange.NormalizeReqAttrs(msg.CreateAskToRemove)
+	askToAdd, err := exchange.NormalizeReqAttrs(msg.CreateAskToAdd)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	bidToRemove, _ := exchange.NormalizeReqAttrs(msg.CreateBidToRemove)
+	bidToAdd, err := exchange.NormalizeReqAttrs(msg.CreateBidToAdd)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+
+	if err = updateReqAttrsAsk(store, marketID, askToRemove, askToAdd); err != nil {
+		errs = append(errs, err)
+	}
+	if err = updateReqAttrsBid(store, marketID, bidToRemove, bidToAdd); err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+
+	k.emitEvent(ctx, exchange.NewEventMarketReqAttrUpdated(marketID, admin))
+	return nil
+}
+
+// getMarketAccountByAddr gets a market's account given its address.
+// This is for when you've already called exchange.GetMarketAddress(marketID) and need it for other things too.
+func (k Keeper) getMarketAccountByAddr(ctx sdk.Context, marketAddr sdk.AccAddress) *exchange.MarketAccount {
+	acc := k.accountKeeper.GetAccount(ctx, marketAddr)
+	if acc == nil {
+		return nil
+	}
+	marketAcc, ok := acc.(*exchange.MarketAccount)
+	if !ok {
+		return nil
+	}
+	return marketAcc
+}
+
+// GetMarketAccount gets a market's account from the account module.
+func (k Keeper) GetMarketAccount(ctx sdk.Context, marketID uint32) *exchange.MarketAccount {
+	marketAddr := exchange.GetMarketAddress(marketID)
+	return k.getMarketAccountByAddr(ctx, marketAddr)
+}
+
+// GetMarketDetails gets a market's details.
+func (k Keeper) GetMarketDetails(ctx sdk.Context, marketID uint32) *exchange.MarketDetails {
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return nil
+	}
+	return &marketAcc.MarketDetails
+}
+
+// UpdateMarketDetails updates a market's details. It returns an error if the market account
+// isn't found or if there aren't any changes provided.
+func (k Keeper) UpdateMarketDetails(ctx sdk.Context, marketID uint32, marketDetails exchange.MarketDetails, updatedBy sdk.AccAddress) error {
+	if err := marketDetails.Validate(); err != nil {
+		return err
+	}
+
+	marketAcc := k.GetMarketAccount(ctx, marketID)
+	if marketAcc == nil {
+		return fmt.Errorf("market %d account not found", marketID)
+	}
+
+	if marketAcc.MarketDetails.Equal(marketDetails) {
+		return errors.New("no changes")
+	}
+
+	marketAcc.MarketDetails = marketDetails
+	k.accountKeeper.SetAccount(ctx, marketAcc)
+	k.emitEvent(ctx, exchange.NewEventMarketDetailsUpdated(marketID, updatedBy))
+
+	return nil
+}
+
+// storeMarket writes all the market fields to the state store (except MarketDetails which are in the account).
+func storeMarket(store sdk.KVStore, market exchange.Market) {
+	marketID := market.MarketId
+	setMarketKnown(store, marketID)
+	setCreateAskFlatFees(store, marketID, market.FeeCreateAskFlat)
```

Comment:
Can we store some of these together? Is this being done for optimization purposes?